src/Native/SymSgdNative/SymSgdNative.cpp

```diff
@@ -0,0 +1,440 @@
+	bool needShuffle, bool shouldInitialize, SymSGDState* state)
+{
+	// If this is the first time LearnAll is called, initialize it.
+	if (shouldInitialize)
+		InitializeState(totalNumInstances, instSizes, instIndices, instValues, numFeat, tuneNumLocIter, numLocIter, numThreads, tuneAlpha, alpha, l2Const, state);
+	float& weightScaling = state->WeightScaling;
+
+	float totalAverageLoss = 0.0f; // Reserved for total loss computation
+	float oldAverageLoss     = INFINITY; 
+	float olderAverageLoss   = INFINITY;
+	float oldestAverageLoss  = INFINITY;
+	float totalOverallAverageLoss = INFINITY;
+
+	float adjustedAlpha = alpha;
+	// Check if totalNumInstances is too small to run in parallel
+	if (numThreads == 1 || totalNumInstances < numThreads) {
+		// For i=[0..totalNumInstances-1], (curPermMultiplier*i) % totalNumInstances always creates a pseduo random permutation
+		int64_t curPermMultiplier = (VERYLARGEPRIME % totalNumInstances);
+		// In the sequential case, just apply normal SGD
+		for (int i = 0; i < numPasses; i++) {
+			for (int j = 0; j < totalNumInstances; j++) {
+				int64_t index = j;
+				if (needShuffle)
+					index = (((int64_t)index * (int64_t)curPermMultiplier) % (int64_t)totalNumInstances);
+				// alpha decays with the square root of number of instances processed.
+				float thisAlpha = adjustedAlpha / (float)sqrt(1 + state->PassIteration * totalNumInstances + j);
+				LearnInstance(instSizes[index], instIndices[index], instValues[index], labels[index], thisAlpha, l2Const, piw,
+					weightScaling, weightVector, bias);
+				//state->TotalInstancesProcessed++;
+				if (weightScaling < 1e-6) {
+					for (int k = 0; k < numFeat; k++) {
+						weightVector[k] *= weightScaling;
+					}
+					weightScaling = 1.0f;
+				}
+			}
+
+			float averageLoss = 0.0f;
+			// Computing the total loss
+			for (int j = 0; j < totalNumInstances; j++)
+				averageLoss += Loss(instSizes[j], instIndices[j], instValues[j], labels[j], piw, weightScaling, weightVector, bias);
+			averageLoss = averageLoss / (float)totalNumInstances;
+			// If we the loss did not improve the learning rate was high, decay it.
+			if (tuneAlpha && oldAverageLoss - averageLoss < tolerance)
+				adjustedAlpha = adjustedAlpha / 10.0f;
+			float overallAverageLoss = oldestAverageLoss - averageLoss;
+			oldestAverageLoss = olderAverageLoss;
+			olderAverageLoss = oldAverageLoss;
+			oldAverageLoss = averageLoss;
+
+			averageLoss = 0.0f;
+			// Terminate if average loss difference between current model and the model from 3 passes ago is small
+			if (overallAverageLoss < tolerance)
+				break;
+
+			// For shuffling in the next passes, instead of curPermMultiplier, use curPermMultiplier^2 which has exactly the same effect
+			if (needShuffle)
+				curPermMultiplier = (((int64_t)curPermMultiplier * (int64_t)curPermMultiplier) % (int64_t)totalNumInstances);
+			state->PassIteration++;
+		}
+	} else {
+#if defined(USE_OMP) 
```

Comment:
What happens in this case if someone passes in `numThreads > 1`?  Does it just no op?

Reply:
yep. We have defined as part of arguments that multi-threaded is not supported. 

---
In reply to: [206914794](https://github.com/dotnet/machinelearning/pull/624#discussion_r206914794) [](ancestors = 206914794)